examples/platform/esp32/mode-support/static-supported-modes-manager.cpp

```diff
@@ -0,0 +1,179 @@
+                                    CHIP_NO_ERROR,
+                                ModeOptionsProvider(nullptr, nullptr));
+            ESP32Config::Key stMfgCodeKey(ESP32Config::kConfigNamespace_ChipFactory, keyBuf);
+            VerifyOrReturnValue(ESP32Config::ReadConfigValue(stMfgCodeKey, semanticTagMfgCode) == CHIP_NO_ERROR,
+                                ModeOptionsProvider(nullptr, nullptr));
+
+            tag.value             = static_cast<uint16_t>(semanticTagValue);
+            tag.mfgCode           = static_cast<chip::VendorId>(semanticTagMfgCode);
+            semanticTags[stIndex] = static_cast<const SemanticTag>(tag);
+        }
+
+        option.label        = chip::CharSpan::fromCharString(modeLabelList[index].supportedModeLabel);
+        option.mode         = static_cast<uint8_t>(supportedModeMode);
+        option.semanticTags = DataModel::List<const SemanticTag>(semanticTags, semanticTagCount);
+
+        modeOptionStruct[index] = option;
+    }
+    endpointArray[endpointId][0] = modeOptionStruct;
+    endpointArray[endpointId][1] = (modeOptionStruct + supportedModeCount);
+
+    return ModeOptionsProvider(modeOptionStruct, modeOptionStruct + supportedModeCount);
+}
+
+Status StaticSupportedModesManager::getModeOptionByMode(unsigned short endpointId, unsigned char mode,
+                                                        const ModeOptionStructType ** dataPtr) const
+{
+    auto modeOptionsProvider = this->getModeOptionsProvider(endpointId);
+    if (modeOptionsProvider.begin() == nullptr)
+    {
+        return Status::UnsupportedCluster;
+    }
+    auto * begin = modeOptionsProvider.begin();
+    auto * end   = modeOptionsProvider.end();
+
+    for (auto * it = begin; it != end; ++it)
+    {
+        auto & modeOption = *it;
+        if (modeOption.mode == mode)
+        {
+            *dataPtr = &modeOption;
+            return Status::Success;
+        }
+    }
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "Cannot find the mode %u", mode);
+    return Status::InvalidCommand;
+}
+
+const ModeSelect::SupportedModesManager * ModeSelect::getSupportedModesManager()
+{
+    return &StaticSupportedModesManager::instance;
+}
+
+void StaticSupportedModesManager::FreeSupportedModes()
+{
+    for (int i = 0; i < FIXED_ENDPOINT_COUNT; i++)
+    {
+        if (endpointArray[i][0] != nullptr)
+        {
+            auto * begin = static_cast<ModeOptionStructType *>(endpointArray[i][0]);
+            auto * end   = static_cast<ModeOptionStructType *>(endpointArray[i][1]);
+            for (auto * it = begin; it != end; ++it)
+            {
+                auto & modeOption = *it;
+                delete[] modeOption.label.data();
+                delete[] modeOption.semanticTags.data();
+            }
+            delete[] begin;
+        }
+        endpointArray[i][0] = nullptr;
+        endpointArray[i][1] = nullptr;
+    }
+    delete[] modeLabelList;
+    delete[] modeOptionStruct;
+    delete[] semanticTags;
```

Comment:
Isn't this deleting things that already got deleted?  Certainly `semanticTags` and `modeLabelList` got deleted when the things corresponding to the last array entry to be filled in were deleted above.

Again, I would _really_ like to understand how this is being tested.

Reply:
Yes removed